File: app/schemas/finance.py
"""
Pydantic schemas for finance endpoints and services.
"""

from __future__ import annotations

from datetime import datetime
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel, Field, validator


class TransactionRecord(BaseModel):
    id: UUID
    amount: int
    transaction_type: str
    category: str
    description: Optional[str]
    created_at: datetime
    balance_after: Optional[int] = None
    related_goal_id: Optional[UUID] = None
    related_shop_item_id: Optional[UUID] = None


class InventoryEntry(BaseModel):
    item_id: str
    item_name: str
    category: Optional[str]
    quantity: int


class LeaderboardEntry(BaseModel):
    user_id: UUID
    balance: int
    care_score: int
    rank: int


class GoalSummary(BaseModel):
    id: UUID
    name: str
    target_amount: int
    current_amount: int
    status: str
    deadline: Optional[datetime]
    completed_at: Optional[datetime]
    progress_percent: float = Field(..., ge=0, le=100)


class ShopItemEntry(BaseModel):
    id: UUID
    sku: str
    name: str
    category: str
    price: int
    stock: int
    description: Optional[str] = None
    metadata: Optional[dict] = None


class FinanceSummary(BaseModel):
    currency: str
    balance: int
    donation_total: int
    lifetime_earned: int
    lifetime_spent: int
    income_today: int
    expenses_today: int
    budget_warning: Optional[str]
    recommendations: List[str] = Field(default_factory=list)
    notifications: List[str] = Field(default_factory=list)
    daily_allowance_available: bool = False
    allowance_amount: int = 0
    goals: List[GoalSummary] = Field(default_factory=list)
    transactions: List[TransactionRecord] = Field(default_factory=list)
    inventory: List[InventoryEntry] = Field(default_factory=list)
    leaderboard: List[LeaderboardEntry] = Field(default_factory=list)


class PurchaseItem(BaseModel):
    item_id: str = Field(..., min_length=1, max_length=120)
    quantity: int = Field(..., ge=1, le=25)


class PurchaseRequest(BaseModel):
    items: List[PurchaseItem]
    pet_id: Optional[UUID] = None

    @validator("items")
    def validate_items(cls, value: List[PurchaseItem]) -> List[PurchaseItem]:
        if not value:
            raise ValueError("At least one item is required for a purchase.")
        return value


class EarnRequest(BaseModel):
    amount: int = Field(..., ge=0)
    reason: str = Field(..., min_length=3, max_length=120)
    care_score: Optional[int] = Field(default=None, ge=0)


class GoalCreateRequest(BaseModel):
    name: str = Field(..., min_length=3, max_length=120)
    target_amount: int = Field(..., gt=0)
    deadline: Optional[datetime] = None


class GoalContributionRequest(BaseModel):
    amount: int = Field(..., gt=0)


class DonationRequest(BaseModel):
    recipient_id: UUID
    amount: int = Field(..., gt=0)
    message: Optional[str] = Field(default=None, max_length=200)


class FinanceResponse(BaseModel):
    summary: FinanceSummary

File: frontend/src/types/finance.ts
export interface TransactionRecord {
  id: string;
  amount: number;
  transaction_type: 'income' | 'expense';
  category: string;
  description?: string | null;
  created_at: string;
  balance_after?: number | null;
  related_goal_id?: string | null;
  related_shop_item_id?: string | null;
}

export interface InventoryEntry {
  item_id: string;
  item_name: string;
  category?: string | null;
  quantity: number;
  shop_item_id?: string | null;
}

export interface LeaderboardEntry {
  user_id: string;
  balance: number;
  care_score: number;
  rank: number;
}

export interface GoalSummary {
  id: string;
  name: string;
  target_amount: number;
  current_amount: number;
  status: 'active' | 'completed' | 'cancelled';
  deadline?: string | null;
  completed_at?: string | null;
  progress_percent: number;
}

export interface ShopItemEntry {
  id: string;
  sku: string;
  name: string;
  category: string;
  price: number;
  stock: number;
  description?: string | null;
  metadata?: Record<string, unknown> | null;
}

export interface FinanceSummary {
  currency: string;
  balance: number;
  donation_total: number;
  lifetime_earned: number;
  lifetime_spent: number;
  income_today: number;
  expenses_today: number;
  budget_warning?: string | null;
  recommendations: string[];
  notifications: string[];
  daily_allowance_available: boolean;
  allowance_amount: number;
  goals: GoalSummary[];
  transactions: TransactionRecord[];
  inventory: InventoryEntry[];
  leaderboard: LeaderboardEntry[];
}

export interface FinanceResponse {
  summary: FinanceSummary;
}

export interface PurchaseLineItemPayload {
  item_id: string;
  quantity: number;
}

export interface PurchaseRequestPayload {
  items: PurchaseLineItemPayload[];
  pet_id?: string | null;
}

export interface EarnRequestPayload {
  amount: number;
  reason: string;
  care_score?: number;
}

export interface GoalCreatePayload {
  name: string;
  target_amount: number;
  deadline?: string | null;
}

export interface GoalContributionPayload {
  amount: number;
}

export interface DonationPayload {
  recipient_id: string;
  amount: number;
  message?: string | null;
}
